common/scpi.c

```diff
@@ -0,0 +1,178 @@
+{
+	return (uint8_t)(uintptr_t)param;
+}
+
+static void
+scpi_send_message(void *param)
+{
+	int      err;
+	uint8_t  chan    = param_to_chan(param);
+	uint64_t timeout = wallclock_read() + SCPI_TIMEOUT;
+
+	/* Wait for previous message to be acknowledged, with a timeout. */
+	while (msgbox_tx_pending(scpi_msgbox, chan)) {
+		if (wallclock_read() >= timeout) {
+			warn("SCPI:%u: Channel busy, retrying send", chan);
+			queue_work(scpi_send_message, param);
+			return;
+		}
+	}
+
+	debug("SCPI:%u: Sending response %08x", chan,
+	      SCPI_MEM_AREA(chan).rx_msg.sender);
+
+	/* Copy the prepared message to the shared memory area. */
+	memcpy(&SCPI_MEM_AREA(chan).tx_msg, &scpi_buffer.tx_msg,
+	       sizeof(struct scpi_msg));
+	/* Mark the buffer as no longer in use. */
+	scpi_buffer_busy = false;
+
+	/* Notify the client that the message has been sent. */
+	if ((err = msgbox_send_msg(scpi_msgbox, chan, SCPI_VIRTUAL_CHAN)))
+		error("SCPI:%u: Error sending response: %d", chan, err);
+}
+
+static void
+scpi_handle_message(void *param)
+{
+	uint8_t chan = param_to_chan(param);
+
+	assert(scpi_buffer_busy);
+
+	/* Create response in tx_msg based on the contents of rx_msg. */
+	scpi_handle_cmd(chan, &scpi_buffer.rx_msg, &scpi_buffer.tx_msg);
+
+	queue_work(scpi_send_message, chan_to_param(chan));
+}
+
+static void
+scpi_create_message(uint8_t chan, uint8_t command)
+{
+	/* Ensure there is space to put the created message. */
+	if (scpi_buffer_busy) {
+		error("SCPI:%u: Cannot create message, buffer is busy", chan);
+		return;
+	}
+	/* Mark the buffer as being in use. */
+	scpi_buffer_busy = true;
+	/* Create the requested message (emulate receiving/handling it). */
+	scpi_buffer.rx_msg.command      = command;
+	scpi_buffer.rx_msg.sender       = 0;
+	scpi_buffer.rx_msg.payload_size = 0;
+
+	queue_work(scpi_handle_message, chan_to_param(chan));
+}
+
+/*
+ * Because the client will regain ownership of the shared memory area once this
+ * function returns, we must do our best to copy its contents to a safe
+ * location before returning.
+ */
+static void
+scpi_receive_message(struct device *dev __unused, uint8_t chan, uint32_t msg)
+{
+	assert(dev == scpi_msgbox);
+	assert(chan == SCPI_NONSEC_CHAN || chan == SCPI_SECURE_CHAN);
+
+	if (msg != SCPI_VIRTUAL_CHAN) {
+		debug("SCPI:%u: Ignoring non-SCPI message 0x%08x", chan, msg);
+		return;
+	}
+
+	debug("SCPI:%u: Received message %08x: %u", chan,
+	      SCPI_MEM_AREA(chan).rx_msg.sender,
+	      SCPI_MEM_AREA(chan).rx_msg.command);
+
+	/* Ensure there is space to put the received message. */
+	if (scpi_buffer_busy) {
+		error("SCPI:%u: Cannot handle message, buffer is busy", chan);
+		return;
+	}
+	/* Mark the buffer as being in use. */
+	scpi_buffer_busy = true;
+	/* Save the received message outside the shared memory area. */
+	memcpy(&scpi_buffer.rx_msg, &SCPI_MEM_AREA(chan).rx_msg,
+	       sizeof(struct scpi_msg));
+
+	queue_work(scpi_handle_message, chan_to_param(chan));
+}
+
+void
```

Comment:
Again, it makes sense to me to return error codes (or add more panic statements) to establish the status of this operation (if the SCPI fails to init, that limits a lot of what we can do).

Reply:
Ok, I'll change the `error()`s to `panic()`s